File: rust/tests/io_expander/mcp23017_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::io_expander::Mcp23017Minimal;
use periph::chips::io_expander::Mcp23017Full;
use embedded_hal::digital::{OutputPin, InputPin};

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x20);

    let dev  = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let chip = Mcp23017Minimal::new(dev, addr).expect("init MCP23017");

    assert!(chip.shadow[0].get() == 0);
    assert!(chip.shadow[1].get() == 0);

    let porta = chip.read_port(0).expect("read_port");
    assert!(porta <= 0xFF);

    let portb = chip.read_port(1).expect("read_port");
    assert!(portb <= 0xFF);

    chip.write_port(0, 0x55).expect("write_port");
    assert_eq!(chip.shadow[0].get(), 0x55);

    chip.write_port(1, 0xAA).expect("write_port");
    assert_eq!(chip.shadow[1].get(), 0xAA);

    let mut p0 = chip.pin(0);
    p0.set_low().expect("set_low");
    assert_eq!(chip.shadow[0].get() & 0x01, 0);

    p0.set_high().expect("set_high");
    assert_eq!(chip.shadow[0].get() & 0x01, 1);

    let v = p0.is_high().expect("is_high");
    assert!(v || !v);  // is_high returns bool; any value is valid

    let p8 = chip.pin(8);
    assert_eq!(p8.n, 8);

    let p15 = chip.pin(15);
    assert_eq!(p15.n, 15);

    // Loopback: PA (outputs) → PB (inputs); PA[n]↔PB[7-n]
    chip.configure_direction(0, 0x00).expect("configure direction A"); // PA all outputs

    chip.write_port(0, 0xAA).expect("write port A"); // PA0=0, avoids contention with PB7 output
    let pb = chip.read_port(1).expect("read port B");
    assert_eq!(pb & 0x7F, 0x55, "loopback 0xAA");

    chip.write_port(0, 0xFE).expect("write port A"); // PA0=0, PA1–PA7=1
    let pb = chip.read_port(1).expect("read port B");
    assert_eq!(pb & 0x7F, 0x7F, "loopback 0xFE");

    chip.write_port(0, 0x00).expect("write port A");
    let pb = chip.read_port(1).expect("read port B");
    assert_eq!(pb & 0x7F, 0x00, "loopback 0x00");

    // Full
    let dev2 = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let full = Mcp23017Full::new(dev2, addr).expect("init MCP23017 full");

    assert!(full.prev[0].get() <= 0xFF);
    assert!(full.prev[1].get() <= 0xFF);

    full.configure_pullup(0, 0x55).expect("configure_pullup");
    full.configure_pullup(1, 0xAA).expect("configure_pullup");

    full.configure_polarity(0, 0x0F).expect("configure_polarity");
    full.configure_polarity(1, 0xF0).expect("configure_polarity");

    let flags = full.read_interrupt_flags(0).expect("read_interrupt_flags");
    assert!(flags <= 0xFF);

    let changed = full.clear_interrupt(0).expect("clear_interrupt");
    assert!(changed <= 0xFF);

    let p1 = full.pin(1);
    let _ = p1;

    println!("===DONE: all checks passed===");
}

File: go/tests/io_expander/mcp23017_test/main.go
//go:build linux && !tinygo

// MCP23017 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the MCP23017 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/io_expander"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x20"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	tr1, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr1.Close()

	chip, err := ioexpander.NewMCP23017Minimal(tr1, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	porta, err := chip.ReadPort(0)
	check("read_porta_range", err == nil && porta <= 0xFF)
	portb, err := chip.ReadPort(1)
	check("read_portb_range", err == nil && portb <= 0xFF)

	if err := chip.WritePort(0, 0x55); err != nil {
		fmt.Fprintln(os.Stderr, "write_porta:", err)
	}
	if err := chip.WritePort(1, 0xAA); err != nil {
		fmt.Fprintln(os.Stderr, "write_portb:", err)
	}
	after, _ := chip.ReadPort(0)
	check("write_porta_0x55", after&0x7F == 0x55)
	after, _ = chip.ReadPort(1)
	check("write_portb_0xAA", after&0x7F == 0xAA)

	p0 := chip.Pin(0)
	if err := p0.Set(false); err != nil {
		fmt.Fprintln(os.Stderr, "set_low:", err)
	}
	if err := p0.Set(true); err != nil {
		fmt.Fprintln(os.Stderr, "set_high:", err)
	}
	_, _ = p0.Get()

	p8 := chip.Pin(8)
	if p8 == (ioexpander.MCP23017Pin{}) {
		fmt.Fprintln(os.Stderr, "pin_8_init:")
	}
	p15 := chip.Pin(15)
	if p15 == (ioexpander.MCP23017Pin{}) {
		fmt.Fprintln(os.Stderr, "pin_15_init:")
	}

	// --- MCP23017Full ---
	tr2, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport full:", err)
		os.Exit(2)
	}
	defer tr2.Close()

	full, err := ioexpander.NewMCP23017Full(tr2, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "new full:", err)
		os.Exit(2)
	}

	if err := full.ConfigurePullup(0, 0x55); err != nil {
		fmt.Fprintln(os.Stderr, "pullup_a:", err)
	}
	if err := full.ConfigurePullup(1, 0xAA); err != nil {
		fmt.Fprintln(os.Stderr, "pullup_b:", err)
	}
	if err := full.ConfigurePolarity(0, 0x0F); err != nil {
		fmt.Fprintln(os.Stderr, "polarity_a:", err)
	}
	if err := full.ConfigurePolarity(1, 0xF0); err != nil {
		fmt.Fprintln(os.Stderr, "polarity_b:", err)
	}
	flags, err := full.ReadInterruptFlags(0)
	check("interrupt_flags_a", err == nil && flags <= 0xFF)
	changed, err := full.ClearInterrupt(0)
	check("clear_interrupt_a", err == nil && changed <= 0xFF)
	fp1 := full.Pin(1)
	_ = fp1

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
